### PR TITLE
pyinfra: 2.9.2 -> 3.0

### PR DESCRIPTION
**Pull request by totoroot**
## Description of changes

Unbreak and update `pyinfra`.

`pyinfra` is currently broken, as building the current version fails with the following error for all Python versions >= 3.12:

```
ModuleNotFoundError: No module named 'distutils'
```

The deprecated usage of `distutils` was already dealt with in `pyinfra` with https://github.com/pyinfra-dev/pyinfra/pull/1096.

~~However, since no 3.x release including these changes is ready, I propose updating the package from `2.9.2` to `3.0.beta2` as this beta release seems already very stable.~~

pyinfra 3.0 just got released so I updated the PR. Find the changelog [here](https://github.com/pyinfra-dev/pyinfra/releases/tag/v3.0).

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/pyinfra/default.nix [hunk 1]
```diff
@@ -6,29 +6,33 @@
...
   setuptools,
+  typeguard,
+  typing-extensions,
 }:
 
 buildPythonPackage rec {
   pname = "pyinfra";
-  version = "2.9.2";
+  version = "3.0";
   format = "setuptools";
```
Comment (mweinelt) [inline pkgs/development/python-modules/pyinfra/default.nix:27]:
```suggestion
  pyproject = true;
```

Migrate from the legacy setuptools format to a PEP517 build.

File: pkgs/development/python-modules/pyinfra/default.nix [hunk 2]
```diff
@@ -6,29 +6,33 @@
...
 
   src = fetchFromGitHub {
     owner = "Fizzadar";
     repo = pname;
     rev = "refs/tags/v${version}";
-    hash = "sha256-lzbFwAg1aLCfBnSnqq4oVteArpkRBa7hU8V3vB5ODa8=";
+    hash = "sha256-0IWfE8SAi9G9gWjRR7CMJ/72V3XW5Ho2acjE/YAooT0=";
   };
 
   propagatedBuildInputs = [
```
Comment (mweinelt) [inline pkgs/development/python-modules/pyinfra/default.nix:38]:
```suggestion
  build-system = [ setuptools ];

  dependencies = [
```

New lingo.

**Review by mweinelt — COMMENTED** [review]
meta.downloadPage is 404.

**Comment by totoroot** [comment]
@mweinelt Applied all your suggestions and squash commits. Thanks!

**Comment by mweinelt** [comment]
Looks like the bloop commit does not belong in here, does it?

**Comment by totoroot** [comment]
You're right. I accidentally authored it during rebase.

File: pkgs/development/python-modules/pyinfra/default.nix [hunk 3]
```diff
@@ -6,43 +6,52 @@
   configparser,
   distro,
   fetchFromGitHub,
+  fetchpatch,
```
Comment (mweinelt) [inline pkgs/development/python-modules/pyinfra/default.nix:9]: unused?

Comment (SuperSandro2000) [inline pkgs/development/python-modules/pyinfra/default.nix:9]:
```suggestion
```

**Comment by fabaff** [comment]
Superseded by #333625.

**Pull request closed without merging**